amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -113,19 +120,31 @@ public TaskResult call() {
                     // This would happen when KinesisDataFetcher(for polling mode) or FanOutRecordsPublisher(for StoS mode) catches ResourceNotFound exception.
                     // In this case, KinesisDataFetcher and FanOutRecordsPublisher will send out SHARD_END signal to trigger a shutdown task with empty list of childShards.
                     // This scenario could happen when customer deletes the stream while leaving the KCL application running.
+                    final Lease currentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
+
                     if (!CollectionUtils.isNullOrEmpty(childShards)) {
                         createLeasesForChildShardsIfNotExist();
-                        updateLeaseWithChildShards();
-                    } else {
-                        log.warn("Shard {} no longer exists. Shutting down consumer with SHARD_END reason without creating leases for child shards.", leaseKeyProvider.apply(shardInfo));
+                        updateLeaseWithChildShards(currentShardLease);
                     }
 
-                    recordProcessorCheckpointer
-                            .sequenceNumberAtShardEnd(recordProcessorCheckpointer.largestPermittedCheckpointValue());
-                    recordProcessorCheckpointer.largestPermittedCheckpointValue(ExtendedSequenceNumber.SHARD_END);
-                    // Call the shardRecordsProcessor to checkpoint with SHARD_END sequence number.
-                    // The shardEnded is implemented by customer. We should validate if the SHARD_END checkpointing is successful after calling shardEnded.
-                    throwOnApplicationException(() -> applicationCheckpointAndVerification(), scope, startTime);
+                    final Lease leaseFromDdb = Optional.ofNullable(leaseCoordinator.leaseRefresher().getLease(leaseKeyProvider.apply(shardInfo)))
+                            .orElseThrow(() -> new IllegalStateException("Lease for shard " + leaseKeyProvider.apply(shardInfo) + " does not exist."));
+
+                    boolean leaseCleanedUp = false;
+                    try {
+                        leaseCleanedUp = leaseCleanupManager.cleanupLease(currentShardLease, shardInfo);
+                    } catch (Exception e) {
+                        log.error("Could not delete lease for shard {}.", leaseKeyProvider.apply(shardInfo), e);
+                    }
+
+                    if (!leaseCleanedUp && !leaseFromDdb.checkpoint().equals(ExtendedSequenceNumber.SHARD_END)) {
```

Comment:
Move the `leaseCleanedUp` as a validation inside this if condition. `if(!shardEnd) {Validate.isTrue(!leaseCleanedUp); ... }`

Reply:
We can sync up on this edge case, but I don't think there's a great way to handle this since the state of the deleted shards is now decoupled from this Shutdown. This is a somewhat narrow edge case though and the worst that happens here is a failed checkpoint.